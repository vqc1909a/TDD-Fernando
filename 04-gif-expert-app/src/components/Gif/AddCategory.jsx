import { useState } from "react"
import PropTypes from 'prop-types'


const DEFAULT_VALUE = "";
 
export const AddCategory = ({handleAddCategory}) => {
  
  const [inputValue, setInputValue] = useState(DEFAULT_VALUE);

  const handleInputChange = (e) => {
    setInputValue(e.target.value);
  }

  const handleSubmit = (e) => { 
    e.preventDefault(); 
    if(inputValue.trim()){
      handleAddCategory(inputValue);
      setInputValue(DEFAULT_VALUE);
    }
  } 
  return (
    <form onSubmit={handleSubmit}>
        <input type="text" placeholder="Buscar gifts" value={inputValue} onChange={handleInputChange}/>
    </form>
  )
}

AddCategory.propTypes = {
  handleAddCategory: PropTypes.func.isRequired
}