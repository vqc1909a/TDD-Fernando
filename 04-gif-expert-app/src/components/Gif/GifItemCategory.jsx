import { PropTypes } from "prop-types"

export const GifItemCategory = ({title, url}) => {
  return (
    <div className="card" aria-label="card">
      <img src={url} alt={title}></img>
      <p>{title}</p>
    </div>
  )
}
GifItemCategory.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
}

/* Tarea

1. AÃ±adir PropTypes a GifItemCategory
  a. title obligatorio
  b. url obligatorio

2. Evaluar el snapshot

*/
