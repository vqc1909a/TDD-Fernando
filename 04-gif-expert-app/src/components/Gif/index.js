//This is a file of exports. It is used to export all the components from the folder Gif, This is also called "barrel" file
export * from "./AddCategory";
export * from "./GifGridCategory";
export * from "./GifItemCategory";

