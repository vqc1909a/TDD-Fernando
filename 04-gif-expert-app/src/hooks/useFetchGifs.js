import { useEffect, useState } from "react";
import { getGifsByCategory } from "../helpers/helperGif";

export const useFetchGifs = (category) => {
  const [gifs, setGifs] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  
  //By the mode strict in the file main.jsx, the console.log show twice as many times as expected. But in production, it will be once time
  // console.log({
  //   gifs
  // })
  
  const getGifs = async () => {
    const gifs = await getGifsByCategory(category);
    setGifs(gifs);
    setIsLoading(false)
  }

  useEffect(() => {
    getGifs();
    //eslint-disable-next-line
  }, [])


  return {
    gifs,
    isLoading
  };

}
