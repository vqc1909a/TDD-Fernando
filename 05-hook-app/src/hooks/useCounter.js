import { useState } from "react"

export const useCounter = (initialValue = 10) => {
  const [counter, setCounter] = useState(initialValue); 

  const handleAddCounter = (value = 1) => {
    setCounter(counter + value);
  }

  const handleResetCounter = (value = initialValue) => {
    setCounter(value);
  }
  const handleSubtractCounter = (value = 1) => {
    if(counter === 0) return;
    setCounter(counter - value);
  }
  return {
    counter,
    handleAddCounter,
    handleResetCounter,
    handleSubtractCounter
  }
}
