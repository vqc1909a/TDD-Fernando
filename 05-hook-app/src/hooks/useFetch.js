import { useEffect, useState } from "react"

export const useFetch = (url) => {
  
  const [state, setState] = useState({
      data: [],
      isLoading: true,
      hasError: false
  })

  const getFetch = async (url) => {
    setState(state => ({ 
        ...state,
        isLoading: true,
    }))

    const resp = await fetch(url)
    const data = await resp.json()
    setState(state => ({
      ...state,       
      data,
      isLoading: false,
      hasError: false 
    }))
  }

  useEffect(() => {
    getFetch(url);
    // eslint-disable-next-line 
  }, [url])
  
  return {
    ...state
  }
}
