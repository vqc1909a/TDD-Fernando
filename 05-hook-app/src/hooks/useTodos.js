import { useCallback, useEffect, useReducer } from "react";
import { todoReducer } from "../08-useReducer/todoReducer";

const init = () => JSON.parse(localStorage.getItem('todos')) || []

export const useTodos = () => {
    //We use the third argument of the useReducer to initialize the state of the reducer    
    const [todos, dispatch] = useReducer(todoReducer, [], init); 

    //The recommended is we have my functions that changed the state in the father component
    const handleAddTodo = useCallback((todo) => {
        dispatch({
            type: "addTodo",
            payload: todo
        })
    }, []);

    const handleDeleteTodo = (id) => {
        dispatch({
            type: "deleteTodo",
            payload: id
        })
    };

    const handleToggleTodo = (id) => {
        dispatch({
            type: "toggleTodo",
            payload: id
        })
    }

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos));
    }, [todos])

    return {
        todos,
        todosCount: todos.length,
        pendingTodosCount: todos.filter(todo => !todo.done).length,
        handleAddTodo,
        handleDeleteTodo,
        handleToggleTodo
    }
}
