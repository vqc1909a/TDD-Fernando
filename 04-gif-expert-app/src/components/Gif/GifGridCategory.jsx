import { PropTypes } from "prop-types"
import { GifItemCategory } from "../../components/Gif";
import { useFetchGifs } from "../../hooks/useFetchGifs";

export const GifGridCategory = ({category}) => {  
  const { gifs, isLoading }  = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>  
      {
        isLoading
        ?
        <p>Loading...</p>
        :
        <div className="card-grid" aria-label="card-grif">
          {gifs.map(image => <GifItemCategory key={image.id} {...image}></GifItemCategory>)}
        </div>
      }
    </>
  )
}

GifGridCategory.propTypes = {
  category: PropTypes.string.isRequired
}

