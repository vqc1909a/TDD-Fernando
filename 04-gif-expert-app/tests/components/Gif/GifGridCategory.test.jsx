import {render, screen} from "@testing-library/react";
import { GifGridCategory } from "../../../src/components/Gif";
import { useFetchGifs } from "../../../src/hooks/useFetchGifs";

// When we make a mock the one function, it will be applicated to each test
jest.mock('../../../src/hooks/useFetchGifs');

describe('tests on component GifGridCategory', () => {
  test('should show the component correctly', () => {

    //For return a value of the mock function, we need to use the function mockReturnValue
    useFetchGifs.mockReturnValue({
      gifs: [],
      isLoading: true
    });

    const category = "One Punch";
    const {container} = render(<GifGridCategory category={category} />)
    expect(container).toMatchSnapshot();
  })

  test('should show the loading correctly', () => {

    useFetchGifs.mockReturnValue({
      gifs: [],
      isLoading: true
    });

    const category = "One Punch";
    render(<GifGridCategory category={category} />)
    expect(screen.getByText('Loading...')).toBeTruthy();

  })

  test('should show gifs when the hook useFetchGifs is loaded with the gifs', () => { 
    const gifs = [
        {
          id: '1',
          title: 'Saitama',
          url: 'https://localhost/saitama.jpg'
        },
        {
          id: '2',
          title: 'Goku',
          url: 'https://localhost/goku.jpg'
        }
    ]
    useFetchGifs.mockReturnValue({
      gifs,
      isLoading: false
    });

    const category = "One Punch";
    render(<GifGridCategory category={category} />)
    expect(screen.getAllByRole('generic', {name: 'card'}).length).toBe(gifs.length);
    expect(screen.getAllByRole('img').length).toBe(gifs.length);

  })
})