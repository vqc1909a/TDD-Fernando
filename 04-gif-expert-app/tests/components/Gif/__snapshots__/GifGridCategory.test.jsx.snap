// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on component GifGridCategory should show the component correctly 1`] = `
<div>
  <h3>
    One Punch
  </h3>
  <p>
    Loading...
  </p>
</div>
`;
